### PR TITLE
emitter/flusher: Initial implementation

### PR DESCRIPTION
**Pull request by sorawit**
Initial implementation of emitter/flusher.

Emitter is an add-on to BandApp to publish state changes to remote Kafka. Currently messages are flushed on every Commit. 

Flusher is a separate Python process that subscribes to messages the push changes to remote databases. It maintains the offset of Kafka messages to make sure no messages are missed.

This is POC implementation. More performance optimizations and features will be added in the future.

### CODE REVIEW

File: chain/emitter/app.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+		if res.IsOK() {
+			extra = app.handleMsg(txHash, msg, logs[idx])
+		}
+		messages = append(messages, JsDict{
+			"msg":   msg,
+			"type":  msg.Type(),
+			"extra": extra,
+		})
+	}
+	txDict["messages"] = messages
```
Comment (Benzbeeb) [inline chain/emitter/app.go:189]: d

File: flusher/flusher/sync.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                value = json.loads(msg.value)
+                if key == "COMMIT":
+                    if value["height"] % commit_interval == 0:
+                        conn.execute(tracking.update().values(kafka_offset=msg.offset))
+                        logger.info(
+                            "Committed at block {} and Kafka offset {}", value["height"], msg.offset
+                        )
+                        break
+                    continue
+                getattr(handler, "handle_" + key.lower())(json.loads(msg.value))
```
Comment (tansawit) [inline flusher/flusher/sync.py]: Why not just use value? No need to do json.loads twice.

**Review by tansawit — CHANGES_REQUESTED** [review]

File: chain/emitter/app.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	invCheckPeriod uint, skipUpgradeHeights map[int64]bool, home string,
+	baseAppOptions ...func(*bam.BaseApp),
+) *App {
+	return &App{
+		BandApp: bandapp.NewBandApp(
+			logger, db, traceStore, loadLatest, invCheckPeriod, skipUpgradeHeights,
+			home, baseAppOptions...,
+		),
+		writer: kafka.NewWriter(kafka.WriterConfig{
+			Brokers:      []string{"localhost:9092"}, // TOD: Remove hardcode
```
Comment (taobun) [inline chain/emitter/app.go]: TODO

File: chain/emitter/handler.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	evMap := make(EvMap)
+	for _, event := range log.Events {
+		for _, kv := range event.Attributes {
+			key := event.Type + "." + kv.Key
+			evMap[key] = append(evMap[key], kv.Value)
+		}
+	}
+	switch msg := msg.(type) {
+	case oracle.MsgRequestData:
+		app.handleMsgRequestData(txHash, msg, evMap, extra)
```
Comment (taobun) [inline chain/emitter/handler.go:21]: Why does this function choose to send extra instead of return extra like handleMsg?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/emitter/handler.go (hunk 4 above)
Comment (sorawit) [reply chain/emitter/handler.go:21]: So we don't need to create a new JsDict instance everywhere.

**Review by taobun — APPROVED** [review]

**Review by Benzbeeb — APPROVED** [review]

**Review by tansawit — APPROVED** [review]

**Pull request merged**